execute/plugin_functions.go

```diff
@@ -365,144 +367,277 @@ func decodeAttributedObservations(
+	}
+
+	if len(result) == 0 {
 		return nil
 	}
+	return result
+}
 
-	return results
+// aggregateMerkleRootObservations groups the observations by merkle root and counts the votes for each.
+// It returns two maps:
+// 1. merkleRootsVotes: a map of merkle root data to the number of votes it received.
+// 2. executedMsgVotes: a map of merkle root data to a map of executed messages and the number of votes they received.
+func aggregateMerkleRootObservations(
+	observations []plugincommon.AttributedObservation[exectypes.Observation],
+) (map[merkleRootData]int, map[merkleRootData]map[cciptypes.SeqNum]int) {
+	merkleRootsVotes := make(map[merkleRootData]int)
+	executedMsgVotes := make(map[merkleRootData]map[cciptypes.SeqNum]int)
+
+	for _, observation := range observations {
+		for sourceChain, merkleRoots := range observation.Observation.CommitReports {
+			for _, merkleRoot := range merkleRoots {
+				data := merkleRootData{
+					SourceChain:         sourceChain,
+					OnRampAddressBase64: base64.StdEncoding.EncodeToString(merkleRoot.OnRampAddress),
+					Timestamp:           merkleRoot.Timestamp,
+					BlockNum:            merkleRoot.BlockNum,
+					MerkleRoot:          merkleRoot.MerkleRoot,
+					SequenceNumberRange: merkleRoot.SequenceNumberRange,
+				}
+
+				merkleRootsVotes[data]++
+
+				if _, ok := executedMsgVotes[data]; !ok {
+					executedMsgVotes[data] = make(map[cciptypes.SeqNum]int)
+				}
+				unqExecutedSeqNums := mapset.NewSet(merkleRoot.ExecutedMessages...).ToSlice()
+				for _, executedSeqNum := range unqExecutedSeqNums {
+					executedMsgVotes[data][executedSeqNum]++
+				}
+			}
+		}
+	}
+
+	return merkleRootsVotes, executedMsgVotes
+}
+
+// merkleRootData is a helper comparable data structure for counting merkle root votes.
+type merkleRootData struct {
+	SourceChain         cciptypes.ChainSelector
+	OnRampAddressBase64 string
+	Timestamp           time.Time
+	BlockNum            uint64
+	MerkleRoot          cciptypes.Bytes32
+	SequenceNumberRange cciptypes.SeqNumRange
 }
 
-func mergeMessageHashes(
+// computeMessageHashesConsensus will iterate over observations of message hashes and come to consensus on them.
+// We select a hash if it has at least f+1 votes. If more than one hashes exist with at
+// least f+1 votes (reaching consensus threshold) then we log an error and skip that specific message.
+func computeMessageHashesConsensus(
 	lggr logger.Logger,
-	aos []plugincommon.AttributedObservation[exectypes.Observation],
+	observations []plugincommon.AttributedObservation[exectypes.Observation],
 	fChain map[cciptypes.ChainSelector]int,
 ) exectypes.MessageHashes {
-	// Single message can transfer multiple tokens, so we need to find consensus on the token level.
-	validators := make(map[cciptypes.ChainSelector]map[cciptypes.SeqNum]consensus.OracleMinObservation[cciptypes.Bytes32])
-	results := make(exectypes.MessageHashes)
+	lggr = logger.With(lggr, "function", "computeMessageHashesConsensus", "fChain", fChain)
```

Comment:
there's already a "caller" field in the logs that tells you the filename/line number